clients/pkg/promtail/client/client.go

```diff
@@ -484,3 +671,116 @@ func (c *client) UnregisterLatencyMetric(labels prometheus.Labels) {
+type clientWAL struct {
+	client     *client
+	tenantWALs map[string]WAL
+	watchers   map[string]stoppable
+}
+
+// newClientWAL creates a new clientWAL.
+func newClientWAL(c *client) clientWAL {
+	return clientWAL{
+		client:     c,
+		tenantWALs: make(map[string]WAL),
+		watchers:   make(map[string]stoppable),
+	}
+}
+
+// getWAL is clientWAL main method. Given a tenantID, it retrieves it's corresponding WAL. If none exists, it creates
+// a new one, and launches a watcher that listens to entries being added, builds batch and sends them when a signaled.
+func (c *clientWAL) getWAL(tenantID string) (WAL, error) {
+	if w, ok := c.tenantWALs[tenantID]; ok {
+		return w, nil
+	}
+	// No WAL exists, create new one and launch watcher. If WAL is disabled, this just returns a no-op implementation.
+	wal, err := newWAL(c.client.logger, c.client.metrics.registerer, c.client.cfg.WAL, c.client.name, tenantID)
+	if err != nil {
+		level.Error(c.client.logger).Log("msg", "could not start WAL", "err", err)
+		return nil, err
+	}
+	// todo: find a cleaner way to do this
+	if c.client.cfg.WAL.Enabled {
+		consumer := newClientConsumer(func(b *batch) error {
+			return c.client.sendBatch(tenantID, b)
+		}, wal, c.client.logger)
+		watcher := NewWALWatcher(wal.Dir(), consumer, c.client.logger)
+		watcher.Start()
+		c.watchers[tenantID] = watcher
+		c.tenantWALs[tenantID] = wal
+	}
+	return wal, nil
+}
+
+func (c *clientWAL) Stop() {
+	for _, watcher := range c.watchers {
+		watcher.Stop()
+	}
+}
+
+type sendBatchFunc func(*batch) error
+
+type SegmentDeleter interface {
+	DeleteSegment(segmentNum int) error
+}
+
+// clientConsumer implements a WatcherConsumer that builds batches with the entries read from the WAL. When the watcher
+// signals that a new segment is found, that means the WAL-writing side decided the entries are enough for a batch to be
+// sent, therefore, the current batch is sent. After the batch is sent successfully, the previous segment can be cleaned
+// up.
+type clientConsumer struct {
+	series         map[uint64]model.LabelSet
+	logger         log.Logger
+	currentBatch   *batch
+	sendBatch      sendBatchFunc
+	segmentDeleter SegmentDeleter
+}
+
+func newClientConsumer(sendBatch sendBatchFunc, segmentDeleter SegmentDeleter, logger log.Logger) *clientConsumer {
+	return &clientConsumer{
+		series:         map[uint64]model.LabelSet{},
+		logger:         logger,
+		currentBatch:   newBatch(0),
+		sendBatch:      sendBatch,
+		segmentDeleter: segmentDeleter,
+	}
+}
+
+func (c *clientConsumer) ConsumeSeries(series record.RefSeries) error {
+	c.series[uint64(series.Ref)] = util.MapToModelLabelSet(series.Labels.Map())
+	return nil
+}
+
+func (c *clientConsumer) ConsumeEntries(samples ingester.RefEntries) error {
+	var entry api.Entry
+	if l, ok := c.series[uint64(samples.Ref)]; ok {
+		entry.Labels = l
+		for _, e := range samples.Entries {
+			entry.Entry = e
+			// Using replay since we know the batch needs to be sent once the segment ends
```

Comment:
So `replay` does basically the same action as `batch.add`, but without doing some checks like maximum batch size, or max stream size, since these have already been performed by the WAL writing side.

Reply:
Can you update the comment to include that detail, so that it's more clear that at the moment we're coupling this side effect of the WAL writing behaviour.

I'm not sure we would want to keep things as coupled as they are currently (side effect of how Cyril/I first implemented the WAL) but it should be clear that this call would need to be changed to `batch.add` when we decouple things in the future.

Reply:
Adding comment, and agree that this should be `batch.add` once we decouple the writing side completely from the batch making.